product_supplierinfo_import/wizards/product_supplierinfo_import.py

```diff
@@ -155,39 +162,113 @@ def _update_create_supplierinfo_data(self, parsed_data):
                 order="date_start desc",
                 limit=1,
             )
-            if previous_product_supplierinfo:
-                previous_product_supplierinfo.date_end = (
-                    self.date_start - relativedelta(days=1)
+            if self.template_id.only_update_existing and previous_supplierinfo:
+                supplier_infos += self._update_existing_supplierinfo(
+                    previous_supplierinfo, row_data
                 )
-            product_supplierinfo = supplier_infos.create(
-                {
-                    "partner_id": self.supplier_id.id,
-                    "product_tmpl_id": product.id,
-                }
+            else:
+                supplier_infos += self._create_new_supplierinfo(
+                    previous_supplierinfo, product, row_data, search_field, search_value
+                )
+        unchanged_supplier_infos = all_supplier_infos - supplier_infos
+        unchanged_supplier_infos.write({"import_status": "unchanged"})
+        return supplier_infos, unchanged_supplier_infos
+
+    def _create_new_product(self, search_value, row_data, search_field):
+        """Creates a new product if allowed."""
+        vendor_product_name_header = self.template_id.template_line_ids.filtered(
+            lambda x: x.field_id.name == "product_name"
+        ).header_name
+        vendor_product_name = row_data.get(
+            vendor_product_name_header,
+            _("%(code)s ((product imported))", code=search_value),
+        )
+        product_data = {
+            "created_from_supplierinfo_import": True,
+            "name": vendor_product_name,
+            "purchase_ok": False,
+            "sale_ok": False,
+            **(
+                {search_field: search_value}
+                if search_field in self.env["product.template"]._fields
+                else {}
+            ),
+        }
+        try:
+            product = self.env["product.template"].create(product_data)
+            product.message_post(
+                body=_("Created from supplier price list import"),
+                type="note",
+            )
+            return product
+        except ValidationError:
+            _logger.warning(
+                f"The product {vendor_product_name}"
+                " could not be created due to a barcode collision."
```

Comment:
Now it could be a barcode collision or something else as the field criteria is flexible